chain-impl-mockchain/src/vote/manager.rs

```diff
@@ -146,17 +183,52 @@ impl ProposalManager {
         &self,
         identifier: account::Identifier,
         payload: ValidatedPayload,
+        token_distribution: &TokenDistribution<TokenIdentifier>,
     ) -> Result<Self, VoteError> {
         // Part of DDoS protection: do not record a new ballot if the account already voted for this
         // proposal. This protects the system from flooding in a system with cheap/free voting
         // transactions.
         let votes_by_voters = self
             .votes_by_voters
-            .insert(identifier, payload)
+            .insert(identifier.clone(), ())
             .map_err(|_| VoteError::AlreadyVoted)?;
+
+        let tally = if let Some(stake) = token_distribution.get_account(&identifier) {
+            match (self.tally.clone(), payload) {
+                (IncrementalTally::Public(mut result), ValidatedPayload::Public(choice)) => {
+                    result.add_vote(choice, stake)?;
+                    IncrementalTally::Public(result)
+                }
+                (
+                    IncrementalTally::Private(mut encrypted_tally),
+                    ValidatedPayload::Private(ballot),
+                ) => {
+                    encrypted_tally.add(&ballot, stake.0);
+                    IncrementalTally::Private(encrypted_tally)
+                }
+                (IncrementalTally::Public(_), ValidatedPayload::Private(_)) => {
+                    return Err(VoteError::InvalidPayloadType {
+                        received: PayloadType::Private,
+                        expected: PayloadType::Public,
+                    })
+                }
+                (IncrementalTally::Private(_), ValidatedPayload::Public(_)) => {
+                    return Err(VoteError::InvalidPayloadType {
+                        received: PayloadType::Public,
+                        expected: PayloadType::Private,
+                    })
+                }
+                (IncrementalTally::Decrypted(_), _) => {
+                    return Err(VoteError::TallyAlreadyDecrypted);
+                }
```

Comment:
Reaching this match arm should be practically impossible. If we try to vote when the tally is already decrypted this should mean we already reached the end of the voting period. I think `unreachable` or `panic` with a comment is a viable option here.

Reply:
Now that you mention it, I just noticed that I accidentally removed the committee date validation checks for the private tally... Or well, rather, they were only done when starting the encrypted tally, and we don't do that anymore, so they need to be done when decrypting the tally, I guess.

Although, this would mean that the decrypted shared needs to be posted during the committee time, which is not the case right now (we don't have an upper bound with the current code in master)

Reply:
Well, unless we only check the lower bound, I guess. Then the `committee_end` would have no meaning for the private tally...

I think this is another thing that we would need to think about if we remove the public variant of the `VoteTally` certificate. If we trigger the governance actions automatically on the voting period end, then I'm not sure what would be the purpose of the committee dates for public voteplans, for example.

Reply:
> Although, this would mean that the decrypted shared needs to be posted during the committee time, which is not the case right now

Mmm, interesting. I think this is an error. From an formal perspective I think it's useful to have an upper bound on committee action (like a timeout) so that we have a clear bound on when a slow committee has to be deemed unresponsive.
This information should be available on chain so that it can be used from smart contracts, scripts, etc.

Reply:
> We can deal with this in another PR though

With what? The smart contracts part?